### ISSUE TITLE
Multiple memory corruption bugs

### ISSUE THREAD

**Issue by cym13** [mananapr/cfiles#3]
# tl;dr: you should protect against buffer overflows.

# How to reproduce

Create a filename with length 255 characters then navigate that directory. The program will segfault.

# Analysis of the issue

I didn't try to pinpoint the exact location of this bug, but the code is written in an unsafe style. Setting arbitrary length of buffers to hold file and directory names is dangerous since a user can use longer filenames.

Also, copying to these buffers without verifying that they fit isn't safe: it ends up writting past the end of the buffer, causing buffer overflows. At best it just doesn't work and the user is forced to use another program, at worst it's a security issue. For example if user A visits a directory with a file created by user B then user B can crash user A's cfiles in a specific way to get code execution, thereby taking over account A.

# Recommendations

- Decide of buffer lengths dynamically depending on the file you find by checking its properties so that it's always at the right size. Don't forget to free them once and only once.

- If you decide to use static buffers, choose their length by looking at the filesystem max filename length, not guessing. Warning, a user routinely uses different types of filesystems even in normal linux usage.

- Use strscpy in order to avoid writting too much to the buffer. If the filename is longer than the buffer, either extend the buffer or deal with the filename one part after the other.

**Comment by mananapr** [comment]
Thanks for the feedback. I'll start working on these issues when I get time

**Comment by mananapr** [comment]
Will using `PATH_MAX` be safe for directory buffers?

**Comment by cym13** [comment]
Yes it would :)

That woud be a bit much since it's meant for the length of a full absolute path and not just one directory's name but it's definitely a safe value and better than an overflow.

**Comment by mananapr** [comment]
In the latest commit efec77f90d9037b418e2f1156b367db52659fd6c, I have replaced `sprintf` with `snprintf` and `strcpy` with `strncpy`. This will prevent buffer overflows. Let me know if this will be enough to close this issue

**Comment by cym13** [comment]
I had a look at the code, there are still many issues. The core reasons are:

- PATH_MAX is meant to hold an absolute path. Any use beyond that (such as holding a shell command containing a path) has no guarantee to be ok. You shouldn't guess the buffer size.

- Using strncpy is difficult to do right since it doesn't always null-terminate and you must indicate how much you want to copy, not the buffer size. I'd recommend always using snprintf instead by replacing `strncpy(out, in, len)` by `snprintf(out, len, "%s", in)`. It's equivalent in functionnality but harder to get wrong.

- Using snprintf or strncpy will prevent the buffer overflow (if used right of course) but you must check that the input wasn't too big for the buffer. Otherwise you will use a truncated input that may cause other important issues.

Here is a rundown of what I noticed while reading, I may have missed some:

- 174: We're never allocating any memory for dir, we're just reusing the one
       that was first allocated for pw_dir. How much is it? Is it safe to
       reuse? This is unsafe.

- 282: Limit should be related to sizeof(temp_dir), not buffer's length

- 371: If strlen(filepath) > 226 then buffer overflow.

- 371: If snprintf is used instead, must still check that the buffer is big
       enough and that there is no overflow.

- 379: No overflow but MIME types can be longer than 20 caracters, causing bugs.
       See https://www.sitepoint.com/mime-types-complete-list/

- 530: If filepath is very long snprintf will truncate the input to fill the
       buffer, so the destination path may not be the right one. That may
       overwrite another file.

- 534: Buffer overflow: if temp_dir is filled, then buf will overflow.

- 546: Same as 530

- 548: Similar to 530

- 687: Similar to 530

- 709: Buffer overflow ; similar to 534

- 743: Similar to 530

- 760: Similar to 530

- 769: Similar to 530

- 792: Similar to 530

**Comment by mananapr** [comment]
Thank you so much for the analysis. So I should use `snprintf()` and pass `NULL` as destination and `0` as size to get the buffer size, make a buffer of that size, use it, and then free it, right?

**Comment by cym13** [comment]
No, what you described would try to write at address 0 which will only crash the program. You should learn about malloc/free.

**Comment by mananapr** [comment]
Checkout the latest commit 6e7f889bc45797993e4f73869e200b120856237 when you get some free time. I have allocated the exact amount of memory needed to all the buffers so there will be no chance of an overflow.
Let me know if I have missed something.

**Comment by cym13** [comment]
That should do it for buffer overflows, I'll start new, more specific issues for the different problems I saw.

**Comment by cym13** [comment]
Actually there is a somewhat huge issue with how you allocate/free memory.

Your strategy so far is (mostly) to have a bunch of global variables. Whenever you want to use them with a new value you first free them, allocate some fresh memory and use them.

The issue happens the first time you free the variable (say `temp_dir`). At that point the variable was never initialized so its value is "random" (remnants of previous memory usage). This is undefined behaviour. There's little margin for exploitation but if an attacker were to be able to control what value will be present in that pointer then the call to free may result in arbitrary code execution (this is similar to a double-free attack if you're interested). Anyway, free is only defined if used on a pointer that was allocated through malloc or NULL.

Solution: initialize all those variables to NULL first.

**Comment by mananapr** [comment]
> Solution: initialize all those variables to NULL first.

 Commit 97662c3 does that. I'll be closing this issue now but feel free to reopen it in case something is missing. Thanks for your help!
